File: 2398-maximum-number-of-robots-within-budget/2398-maximum-number-of-robots-within-budget.py
from sortedcontainers import SortedList

class Solution:
    def maximumRobots(self, chargeTimes: List[int], runningCosts: List[int], budget: int) -> int:
        n = len(chargeTimes)
        j = 0
        ans = 0
        total = 0
        sl = SortedList()
        
        for i in range(n):
            total += runningCosts[i]
            sl.add(chargeTimes[i])
            while sl and sl[-1] + (i - j + 1) * total > budget:
                total -= runningCosts[j]
                sl.remove(chargeTimes[j])
                j += 1
                
            ans = max(ans, i - j + 1)
            
        return ans

File: 2398-maximum-number-of-robots-within-budget/2398-maximum-number-of-robots-within-budget.cpp
using ll = long long;

class Solution {
public:
    int maximumRobots(vector<int>& chargeTimes, vector<int>& runningCosts, long long budget) {
        multiset<ll> ms;
        ll j = 0, s = 0, ans = 0;
        ll n = chargeTimes.size();
        
        for (int i = 0; i < n; ++i) {
            ms.insert(chargeTimes[i]);
            s += runningCosts[i];
            while (j < n and ms.size() and *ms.rbegin() + (i - j + 1)*s > budget) {
                s -= runningCosts[j];
                ms.erase(ms.lower_bound(chargeTimes[j]));
                j++;
            }
            ans = max(ans, i - j + 1);
        }
        
        return ans;
    }
};